jolt-core/src/r1cs/special_polys.rs

```diff
@@ -1,43 +1,499 @@
+        assert!(b.Z.len() >= a.Z.len() && b.Z.len() >= c.Z.len());
+
+        // TODO(sragss): Explain the strategy
+
+        let (b_chunks, mut dense_ranges) = b.chunk_no_split_siblings(n);
+        let highest_non_zero = [&a.Z, &b.Z, &c.Z]
+            .iter()
+            .filter_map(|z| z.last().map(|&(_, index)| index))
+            .max()
+            .unwrap();
+        dense_ranges.last_mut().unwrap().1 = highest_non_zero + 1;
+        assert_eq!(b_chunks.len(), n);
+        assert_eq!(dense_ranges.len(), n);
+
+        // Create chunks of (a, c) which overlap with b's sparse indices
+        let mut a_chunks: Vec<&[(F, usize)]> = vec![&[]; n];
+        let mut c_chunks: Vec<&[(F, usize)]> = vec![&[]; n];
+        let mut a_i = 0;
+        let mut c_i = 0;
+        let span = tracing::span!(tracing::Level::DEBUG, "a_c_chunking");
+        let _enter = span.enter();
+        for (chunk_index, range) in dense_ranges.iter().enumerate().skip(1) {
+            // Find the corresponding a, c chunks
+            let prev_chunk_end = range.0;
+
+            if a_i < a.Z.len() && a.Z[a_i].1 < prev_chunk_end {
+                let a_start = a_i;
+                while a_i < a.Z.len() && a.Z[a_i].1 < prev_chunk_end {
+                    a_i += 1;
+                }
+
+                a_chunks[chunk_index - 1] = &a.Z[a_start..a_i];
+            }
+
+            if c_i < c.Z.len() && c.Z[c_i].1 < prev_chunk_end {
+                let c_start = c_i;
+                while c_i < c.Z.len() && c.Z[c_i].1 < prev_chunk_end {
+                    c_i += 1;
+                }
+
+                c_chunks[chunk_index - 1] = &c.Z[c_start..c_i];
+            }
+        }
+        drop(_enter);
+        a_chunks[n - 1] = &a.Z[a_i..];
+        c_chunks[n - 1] = &c.Z[c_i..];
+
+        #[cfg(test)]
+        {
+            assert_eq!(a_chunks.concat(), a.Z);
+            assert_eq!(b_chunks.concat(), b.Z);
+            assert_eq!(c_chunks.concat(), c.Z);
+        }
+
+        let mut iterators: Vec<SparseTripleIterator<'a, F>> = Vec::with_capacity(n);
+        for (((a_chunk, b_chunk), c_chunk), range) in a_chunks
+            .iter()
+            .zip(b_chunks.iter())
+            .zip(c_chunks.iter())
+            .zip(dense_ranges.iter())
+        {
+            #[cfg(test)]
+            for chunk in &[a_chunk, b_chunk, c_chunk] {
+                for (_, index) in chunk.iter() {
+                    assert!(*index >= range.0 && *index <= range.1);
+                }
+            }
+
+            let iter = SparseTripleIterator {
+                dense_index: range.0,
+                end_index: range.1,
+                a: a_chunk,
+                b: b_chunk,
+                c: c_chunk,
+            };
+            iterators.push(iter);
+        }
+
+        iterators
+    }
+
+    pub fn has_next(&self) -> bool {
+        self.dense_index < self.end_index
+    }
+
+    pub fn next_pairs(&mut self) -> (usize, F, F, F, F, F, F) {
```

Comment:
I wanted to avoid using the return type `Option<Self::Item>` as we have enough conditional / branching logic in the hot loop as is. But I'm not certain it would actually close a slowdown. `next_pairs()` also seemed clearer to me. Given the obvious thing for the `SparseTripleIterator.next()` to return would be 3 `F` elements.